Register WPE_SETTING_DRM_SCALE and provide a sane default

aperezdc opened the pull request:
#### f2bd2968f34cb46e63e17e3cfd2151882a177046
<pre>
Register WPE_SETTING_DRM_SCALE and provide a sane default
<a href="https://bugs.webkit.org/show_bug.cgi?id=289230">https://bugs.webkit.org/show_bug.cgi?id=289230</a>

Reviewed by Carlos Garcia Campos.

* Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
(wpeDisplayDRMConstructed): Added, registers the WPE_SETTING_DRM_SCALE
setting.
(wpeDisplayDRMSetup): Improve the logic that parses the WPE_DRM_SCALE
environment variable, checking for parsing errors and the value of the
variable. If the value from the environment cannot be used, call
wpeScreenDRMGuessScale() to find a reasonable default for the screen.
While at it, bounds-check values possibly set by the application and
avoid using them if too small or too big.
(wpe_display_drm_class_init):
* Source/WebKit/WPEPlatform/wpe/drm/WPEScreenDRM.cpp:
(wpeScreenDRMGuessScale): Added, switches from 1.0 scaling to 2.0 if the
screen is at least 1200px tall and has a density of 192dpi. The logic
could use more finesse, choosing fractional values in between but works
well enough as a first approach.
* Source/WebKit/WPEPlatform/wpe/drm/WPEScreenDRMPrivate.h:

Canonical link: <a href="https://commits.webkit.org/291757@main">https://commits.webkit.org/291757@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ddb3317d5576b9d2ed51d950de6cf9c27ac10b3c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93860 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13443 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3175 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98866 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44386 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13738 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21876 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71628 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28986 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96862 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10202 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84805 "Found 1 new API test failure: TestWebKitAPI.WKWebExtensionAPIDevTools.MessagePassingFromPanelToBackground (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51963 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9886 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43702 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80151 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2518 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/100994 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20912 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15254 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80643 "layout-tests (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21164 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80763 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79989 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24538 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1902 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14069 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15067 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20896 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26074 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20583 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24043 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22324 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
```diff
@@ -307,19 +318,45 @@ static gboolean wpeDisplayDRMSetup(WPEDisplayDRM* displayDRM, const char* device
+        if (parsedLength == trimmedScaleString.length() && scaleIsInBounds(scale))
+            scaleFromEnvironment = scale;
+        else
+            g_warning("Invalid WPE_DRM_SCALE='%*s' value, or out of bounds.", int(scaleString.span8().size()), scaleString.span8().data());
```
```suggestion
            g_warning("Invalid WPE_DRM_SCALE='%*s' value, or out of bounds.", static_cast<int>(scaleString.span8().size()), scaleString.span8().data());
```

aperezdc replied:
Ah, this slipped. Sometimes to type less I just to `T()` for integral types, or even (gasp!) C casts while working on a patch. I'll change it to a `static_cast` before landing 👍🏼

aperezdc replied:
Done.

spenap commented on Source/WebKit/WPEPlatform/wpe/drm/WPEScreenDRM.cpp:
```diff
@@ -113,6 +113,33 @@ const WPE::DRM::Crtc wpeScreenDRMGetCrtc(WPEScreenDRM* screen)
+        return 1.0;
+
+    const auto horizontalDPI = float(pixelWidth) / (physicalWidth / mmPerInch);
+    const auto verticalDPI = float(pixelHeight) / (physicalHeight / mmPerInch);
```
```suggestion
    const auto horizontalDPI = static_cast<float>(pixelWidth) / (physicalWidth / mmPerInch);
    const auto verticalDPI = static_cast<float>(pixelHeight) / (physicalHeight / mmPerInch);
```

aperezdc replied:
Done.

spenap reviewed (commented):
Or maybe `{}`? See https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es64-use-the-tenotation-for-construction

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
```diff
@@ -307,19 +318,45 @@ static gboolean wpeDisplayDRMSetup(WPEDisplayDRM* displayDRM, const char* device
+        // The value has been set by the application but is unusable. Reset
+        // the application-set value to force using the platform-set default.
+        RELEASE_ASSERT(wpe_settings_set_value(wpe_display_get_settings(WPE_DISPLAY(displayDRM)), WPE_SETTING_DRM_SCALE, nullptr, WPE_SETTINGS_SOURCE_APPLICATION, nullptr));
+        scale = wpe_settings_get_double(wpe_display_get_settings(WPE_DISPLAY(displayDRM)), WPE_SETTING_DRM_SCALE, &settingsError.outPtr());
```
I don't think we should do that, platforms should never set settings with application source. If the value we get is out of bounds, we should just ignore it and use default or computed one.

aperezdc replied:
Sure, I'll remove it.

aperezdc replied:
Done, removed.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
```diff
@@ -307,19 +318,45 @@ static gboolean wpeDisplayDRMSetup(WPEDisplayDRM* displayDRM, const char* device
+    };
+
+    std::optional<double> scaleFromEnvironment;
+    if (const auto scaleString = StringView::fromLatin1(getenv("WPE_DRM_SCALE"))) {
+        RELEASE_ASSERT(scaleString.is8Bit());
+        auto trimmedScaleString = scaleString.trim(isASCIIWhitespace<LChar>);
+        size_t parsedLength = 0;
+        auto scale = parseDouble(trimmedScaleString, parsedLength);
+        if (parsedLength == trimmedScaleString.length() && scaleIsInBounds(scale))
+            scaleFromEnvironment = scale;
```
Is all this simpler than `g_ascii_strtod`?  Or is it because it's safer?

aperezdc replied:
Easier to follow when not using raw pointers, and harder to make mistakes (which means: a tad safer). Also, if we add to the C-like version the missing whitespace trimming and checking for a successful parse, it turns out to be about the same length:

```cpp
if (const char* scaleString = getenv("WPE_DRM_SCALE")) {
    scaleString = g_strstrip(scaleString);
    char *endptr = NULL;
    auto scale = g_ascii_strtod(scaleString, &endptr);
    if (*endptr == '\0' && (errno != ERANGE) && scaleIsInBounds(scale))
        scaleFromEnvironment = scale;
}
```

In light of this, I think the version proposed in the patch is nicer.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
```diff
@@ -66,6 +69,14 @@ WEBKIT_DEFINE_FINAL_TYPE_WITH_CODE(WPEDisplayDRM, wpe_display_drm, WPE_TYPE_DISP
+{
+    G_OBJECT_CLASS(wpe_display_drm_parent_class)->constructed(object);
+
+    auto* settings = wpe_display_get_settings(WPE_DISPLAY(object));
+    RELEASE_ASSERT(wpe_settings_register(settings, WPE_SETTING_DRM_SCALE, G_VARIANT_TYPE_DOUBLE, g_variant_new_double(1.0), nullptr));
```
Oh, we forgot to register the setting when we introduced it! 🤦‍♂️

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/drm/WPEScreenDRM.cpp:
```diff
@@ -113,6 +113,33 @@ const WPE::DRM::Crtc wpeScreenDRMGetCrtc(WPEScreenDRM* screen)
 
+double wpeScreenDRMGuessScale(WPEScreenDRM* screen)
+{
+    static constexpr uint32_t minimum2xScaleDPI = 2 * 96;
+    static constexpr uint32_t minimum2xPixelHeight = 1200;
+    static constexpr float mmPerInch = 25.4;
```
I would add a comment here explaining the logic we are following to decide the scale.

aperezdc replied:
Will do 👍🏼

aperezdc replied:
Comment added.

aperezdc committed:
Register WPE_SETTING_DRM_SCALE and provide a sane default
https://bugs.webkit.org/show_bug.cgi?id=289230

Reviewed by Carlos Garcia Campos.

* Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
(wpeDisplayDRMConstructed): Added, registers the WPE_SETTING_DRM_SCALE
setting.
(wpeDisplayDRMSetup): Improve the logic that parses the WPE_DRM_SCALE
environment variable, checking for parsing errors and the value of the
variable. If the value from the environment cannot be used, call
wpeScreenDRMGuessScale() to find a reasonable default for the screen.
While at it, bounds-check values possibly set by the application and
avoid using them if too small or too big.
(wpe_display_drm_class_init):
* Source/WebKit/WPEPlatform/wpe/drm/WPEScreenDRM.cpp:
(wpeScreenDRMGuessScale): Added, switches from 1.0 scaling to 2.0 if the
screen is at least 1200px tall and has a density of 192dpi. The logic
could use more finesse, choosing fractional values in between but works
well enough as a first approach.
* Source/WebKit/WPEPlatform/wpe/drm/WPEScreenDRMPrivate.h:

Canonical link: https://commits.webkit.org/291757@main
